bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java

```diff
@@ -1228,82 +1230,174 @@ private static boolean checkIsArrayType(BType sourceType, BArrayType targetType,
             return true;
         }
 
-        if (sourceType.getTag() != TypeTags.ARRAY_TAG && sourceType.getTag() != TypeTags.TUPLE_TAG) {
+        if (sourceTypeTag != TypeTags.ARRAY_TAG && sourceTypeTag != TypeTags.TUPLE_TAG) {
             return false;
         }
 
+        BType targetElementType = targetType.getElementType();
+        int targetElementTypeTag = targetElementType.getTag();
+
         BArrayType sourceArrayType;
-        if (sourceType.getTag() == TypeTags.ARRAY_TAG) {
+        if (sourceTypeTag == TypeTags.ARRAY_TAG) {
             sourceArrayType = (BArrayType) sourceType;
         } else {
             BTupleType sourceTupleType = (BTupleType) sourceType;
             Set<BType> tupleTypes = new HashSet<>(sourceTupleType.getTupleTypes());
-            if (sourceTupleType.getRestType() != null) {
-                tupleTypes.add(sourceTupleType.getRestType());
+            BType sourceRestType = sourceTupleType.getRestType();
+
+            if (sourceRestType != null) {
+                tupleTypes.add(sourceRestType);
```

Comment:
Instead of adding the rest-descriptor to tupleTypes can we handle it separately?

For example, 
- if the target array is closed, the source cannot have a rest-descriptor (return false directly). If there is no rest-descriptor it may be valid, so check each member against the target array's element type - we seem to be doing this already
- if the target array is open, check member types against element type, and then check the rest-descriptor against the target element type separately - add another check to the else block below